File: glug-gui-components/src/test/java/glug/gui/timelinecursor/TimelineCursorTest.java
package glug.gui.timelinecursor;

import glug.gui.timelinecursor.TimelineCursor.CursorPositionChanged;
import glug.model.time.LogInstant;
import glug.model.time.LogInterval;
import org.junit.Test;

import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.nullValue;
import static org.joda.time.Duration.standardSeconds;


public class TimelineCursorTest {

    @Test
    public void shouldFireAChangeEventThatContainsCorrectOldAndNewState() {
        TimelineCursor cursor = new TimelineCursor();
        LogInterval intervalSelected = new LogInterval(standardSeconds(5), new LogInstant(8000,0));
        cursor.setDot(intervalSelected.getStart());
        cursor.moveDot(intervalSelected.getEnd());

        ChangeListenerStub changeListenerStub = new ChangeListenerStub();
        cursor.addChangeListener(changeListenerStub);

        LogInstant newDot = new LogInstant(18000, 0);
        cursor.setDot(newDot);

        TimelineCursor.CursorPositionChanged o = (CursorPositionChanged) changeListenerStub.getChangeEvent().getSource();
        assertThat(o.getOldState().getSelectedInterval(), equalTo(intervalSelected));
        assertThat(o.getNewState().getSelectedInterval(), nullValue());
        assertThat(o.getNewState().getDot(), equalTo(newDot));

    }

    static class ChangeListenerStub implements ChangeListener {

        private ChangeEvent changeEvent;

        @Override
        public void stateChanged(ChangeEvent changeEvent) {
            this.changeEvent = changeEvent;
        }

        public ChangeEvent getChangeEvent() {
            return changeEvent;
        }

    }
}

File: glug-gui-components/src/test/scala/glug/gui/timelinecursor/TimelineCursorTest.scala
package glug.gui.timelinecursor

import javax.swing.event.{ChangeEvent, ChangeListener}

import glug.model.time.{LogInstant, LogInterval}
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.{equalTo, nullValue}
import org.joda.time.Duration.standardSeconds
import org.junit.Test

object TimelineCursorTest {

  private[timelinecursor] class ChangeListenerStub extends ChangeListener {
    private var changeEvent: ChangeEvent = null

    override def stateChanged(changeEvent: ChangeEvent) {
      this.changeEvent = changeEvent
    }

    def getChangeEvent: ChangeEvent = changeEvent
  }

}

class TimelineCursorTest {
  @Test def shouldFireAChangeEventThatContainsCorrectOldAndNewState(): Unit = {
    val cursor = new TimelineCursor
    val intervalSelected = LogInterval(standardSeconds(5), LogInstant(8000, 0))
    cursor.setDot(intervalSelected.getStart)
    cursor.moveDot(intervalSelected.getEnd)
    val changeListenerStub = new TimelineCursorTest.ChangeListenerStub
    cursor.addChangeListener(changeListenerStub)
    val newDot = new LogInstant(18000, 0)
    cursor.setDot(newDot)
    val o = changeListenerStub.getChangeEvent.getSource.asInstanceOf[TimelineCursor.CursorPositionChanged]
    assertThat(o.getOldState.getSelectedInterval, equalTo(intervalSelected))
    assertThat(o.getNewState.getSelectedInterval, nullValue)
    assertThat(o.getNewState.getDot, equalTo(newDot))
  }
}